### ISSUE TITLE
infinite loop in hcxpcaptool

### ISSUE THREAD

**Issue by hops** [ZerBea/hcxtools#98]
I've been playing around with honggfuzz and found a testcase causing hcxpcaptool to be stuck in an infinite loop.

Steps to reproduce:
```
$ cat loop.b64
Cg0NCmAAAABNPCsaAQAAAP//////////AgAGAHg4Nl82NAAAAwAYAExpbnV4IDUuMS41LWFyY2gx
LTItQVJDSAQAEQBoY3hkdW1wdG9vbCA1LjEuNQAAAAAAAABgAAAABQAAAAAAAAAAYAAAAAUAAAAA
AAAAAGAAAAAFAAAAKAAAAH8AAAD//wAAAgAJAHdscDBzMjB1MQAAAAAAAA==

$ base64 -d loop.b64 > loop.bin
$ ./hcxpcaptool -o /dev/null loop.bin

reading from loop.bin
```
This uses up 100% CPU and never ends.

**Comment by ZerBea** [comment]
That is really an ugly bug. Thanks for reporting it!
Should be fixed by commit:
https://github.com/ZerBea/hcxtools/commit/5d90b751beda75d450f8869426153699150cd007

$ hcxpcaptool -o test.hccapx loop.bin
reading from loop.bin
invalid blocktype length detected
summary:                                        
file name........................: loop.bin
file type........................: pcapng 1.0
file hardware information........: x86_64
file os information..............: Linux 5.1.5-arch1-2-ARCH
file application information.....: hcxdumptool 5.1.5
network type.....................: DLT_NULL (0)
endianness.......................: little endian
read errors......................: yes
packets inside...................: 0
skipped packets.(damaged)........: 0
packets with GPS data............: 0
packets with FCS.................: 0

Please test and close issue, if it works for you.

**Comment by ZerBea** [comment]
Maybe afl check was too short.
https://www.archlinux.org/packages/community/x86_64/afl/
Next time I'll run it longer.

**Comment by hops** [comment]
I managed to add a [libfuzzer](https://github.com/google/honggfuzz/blob/master/docs/FeedbackDrivenFuzzing.md#llvm-style-llvmfuzzertestoneinput) test.
After I clean up my fuzzing code a bit I'll push it somewhere to look at and maybe integrate it into [oss-fuzz](https://github.com/google/oss-fuzz)
hongfuzz was able to  test ~30k cases per second on a i7-5600U notebook CPU.
This found another infinite loop and a crash due to a stack buffer overflow.
```
$ cat /tmp/loop.b64 
Cg0NCmAAAABNPCsaAQAAAP/////8////AgAGAHg4Nl83NAAAAwAYAExpZHV4IDUuMQAAAAJyY2gx
LTIGAAAASAQAEQBoY3hrdW1wdG9vYCA1LjEuNQAAAAAAAABgAAAABgAAABwAAAB/AAYAZAAAAAAA
AP86AAAAhAEAAAIAAAAPAAAAAAAAAGAAAAAGAAAAHAAAAH8ABgBkAKfy/////zgAAACEAQAAAgAA
AGgfAC00NDY1Mxb8+r8ABgAAAAYA

$ base64 -d loop.b64 > loop.bin
$ ./hcxpcaptool -o /dev/null /tmp/loop.bin 

reading from loop.bin
1300000 packets processed - be patient!
^C
```

```
$ cat crash.b64 
Cg0NCmAAAABNPNQaAQABAAAAAAAAgN//AgAGAHg4Nl82NgAAAwAXAExpbnV4IDUuMS41LWFyY2gx
rTItQVJDSJzyQABoY3hkdW1wdG9vbCA1LjEuNQAAAC02MDA0MDY1NdMBAAAAAAAAgN//AgAGAHg4
Nl82NgAAAwAXAExpbnV4IDUuMS41LWFyY2gxrTItQVJDSJzyQABoY3hkdW1wdG9vbCA1LjEuNQAA
AC02MDA0MDY1NdMBAAAAAAA=

$ base64 -d crash.b64 > crash.bin
$ ./hcxpcaptool -o /dev/null crash.bin 

reading from crash.bin
*** stack smashing detected ***: <unknown> terminated
Aborted (core dumped)
```

**Comment by RealEnder** [comment]
oss-fuzz integration will be awesome. I did several rounds of fuzzing with afl in the past, which was useful, but doing it on scale will definitely be great improvement!

**Comment by ZerBea** [comment]
Again thanks for reporting that issues. Fixed by this commit:
https://github.com/ZerBea/hcxtools/commit/d988214f8d819e03021e271840cfa7f323eaa879

For the first one:
$ hcxpcaptool -V loop.bin

reading from loop.bin
failed to read packet 0          
summary:                                        
file name........................: loop.bin
file type........................: pcapng 1.0
file hardware information........: x86_74
file os information..............: Lidux 5.1
file application information.....: hcxkumptoo` 5.1.5
network type.....................: DLT_NULL (0)
endianness.......................: little endian
read errors......................: yes
packets inside...................: 0
skipped packets.(damaged)........: 0
packets with GPS data............: 0
packets with FCS.................: 0

and the second one:
$ hcxpcaptool -V loop.bin

reading from loop.bin
failed to read pcapng options
                                                
summary:                                        
file name........................: loop.bin
file type........................: pcapng 1.1
file hardware information........: x86_66
file os information..............: Linux 5.1.5-arch1�2-ARCH
file application information.....: unknown
network type.....................: DLT_NULL (0)
endianness.......................: little endian
read errors......................: yes
packets inside...................: 0
skipped packets.(damaged)........: 0
packets with GPS data............: 0
packets with FCS.................: 0

Both were very hard to find!

**Comment by ZerBea** [comment]
Please let me know, when your testcase finished.  If no more show-stopper problems are found, I'll move to 5.1.7

**Comment by hops** [comment]
I found another buffer overflow and another loop.
loop:
```
Cg0NChEAAAD8if9zjOZF01oCAAAADQAAAAAAAACc/A3nZS7TWgIAAAD8nDg4NTg2MjgxMDI3MDQx
NTYxNTYxsLPiz3xHRu83ea6ccf7GILQs7Dt1lc1LPWiKXlQojfy7K+xgI2/vFmK2sDETi8AnYMeu
nPwA
```
buffer overflow (only triggers with Address Sanitizer compile with:
`clang -O3 -Wall -Wextra -ggdb -std=gnu99 -fsanitize=address  -MMD -MF .deps/hcxpcaptool.d -o hcxpcaptool hcxpcaptool.c -lz -lcrypto`
```
Cg0NCiCUFpX2HYxH91PKbJ57AAFt2hzHnPJcAC8XnnR8/tbfAAD/Cg0NChgABAAALxcAAAAN1v8A
AP8KDQ0KGAAEAAAAAP8AQNb/AAD/Cg0NLxgAAAAvvxivXy00MjY5
```
```
./hcxpcaptool -o /dev/null fuzz-testcases/crash.bin 

reading from crash.bin
=================================================================
==15214==ERROR: AddressSanitizer: stack-buffer-overflow on address 0x7ffc8dafb998 at pc 0x56299870c227 bp 0x7ffc8dafb950 sp 0x7ffc8dafb100
WRITE of size 68 at 0x7ffc8dafb998 thread T0
    #0 0x56299870c226 in __interceptor_read.part.51 (/home/hops/src/hcxtools/hcxpcaptool+0x97226)
    #1 0x5629987e0da3 in pcapngoptionwalk /home/hops/src/hcxtools/hcxpcaptool.c:5297:9
    #2 0x5629987e17ed in processpcapng /home/hops/src/hcxtools/hcxpcaptool.c:5426:7
    #3 0x5629987e5f5e in processcapfile /home/hops/src/hcxtools/hcxpcaptool.c:6091:2
    #4 0x5629987e90bc in main /home/hops/src/hcxtools/hcxpcaptool.c:6577:2
    #5 0x7f1e57633ce2 in __libc_start_main (/usr/lib/libc.so.6+0x23ce2)
    #6 0x56299869f24d in _start (/home/hops/src/hcxtools/hcxpcaptool+0x2a24d)

Address 0x7ffc8dafb998 is located in stack of thread T0 at offset 56 in frame
    #0 0x5629987e08af in pcapngoptionwalk /home/hops/src/hcxtools/hcxpcaptool.c:5178

  This frame has 3 object(s):
    [32, 36) 'opthdr' (line 5182)
    [48, 56) 'filereplaycound' (line 5192)
    [80, 112) 'filenonce299' (line 5193) <== Memory access at offset 56 partially underflows this variable
HINT: this may be a false positive if your program uses some custom stack unwind mechanism, swapcontext or vfork
      (longjmp and C++ exceptions *are* supported)
SUMMARY: AddressSanitizer: stack-buffer-overflow (/home/hops/src/hcxtools/hcxpcaptool+0x97226) in __interceptor_read.part.51
Shadow bytes around the buggy address:
  0x100011b576e0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x100011b576f0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x100011b57700: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x100011b57710: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x100011b57720: 00 00 00 00 00 00 00 00 00 00 00 00 f1 f1 f1 f1
=>0x100011b57730: 04 f2 00[f2]f2 f2 00 00 00 00 f3 f3 f3 f3 f3 f3
  0x100011b57740: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x100011b57750: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x100011b57760: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x100011b57770: 00 00 00 00 00 00 00 00 f1 f1 f1 f1 00 f2 f2 f2
  0x100011b57780: 00 00 f2 f2 00 f2 f2 f2 00 00 04 f2 f2 f2 f2 f2
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07 
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
  Shadow gap:              cc
==15214==ABORTING
```

**Comment by ZerBea** [comment]
Great, thanks again.
Fixed by commit:
https://github.com/ZerBea/hcxtools/commit/4c8accaca337a1e9a59f0e98c2c2e822ec34ac73

$ hcxpcaptool -V loop.bin

reading from loop.bin
unknown blocktype -751933977 after packet 1
summary:                                        
file name........................: loop.bin
file type........................: pcapng 59020.54085
file hardware information........: unknown
file os information..............: unknown
file application information.....: unknown
network type.....................: DLT_NULL (0)
endianness.......................: little endian
read errors......................: yes
packets inside...................: 1
skipped packets.(damaged)........: 0
packets with GPS data............: 0
packets with FCS.................: 0
warning..........................: zero value timestamps detected - this prevents EAPOL timeout calculation


$ hcxpcaptool -V loop.bin
reading from loop.bin
block length greater than file size
summary:                                        
--------
file name........................: loop.bin
file type........................: pcapng 21495.27850
file hardware information........: unknown
file os information..............: unknown
file application information.....: unknown
network type.....................: DLT_NULL (0)
endianness.......................: little endian
read errors......................: yes
packets inside...................: 0
skipped packets.(damaged)........: 0
packets with GPS data............: 0
packets with FCS.................: 0

**Comment by hops** [comment]
Thank you for fixing these bugs so quickly!

And another crash only triggered if compiled with Address Sanitizer:
```
1MOyoRAAoX4tHDAp990PPgAAAAAAAAAAAAAAAAAfGx8eAAAAACFDAAIAAABAppCQABQPAAERABIA
QKSQkAAUDwABEQCt2oFzUWkj0FsBEQCt2oFzUWkj0FsBEQCt2oEA
```

```
$ ./hcxpcaptool -o /dev/null fuzz-testcases/crash.bin 

reading from crash.bin
=================================================================
==12587==ERROR: AddressSanitizer: stack-buffer-overflow on address 0x7ffecf399241 at pc 0x5649b384aa35 bp 0x7ffecf319010 sp 0x7ffecf319008
READ of size 1 at 0x7ffecf399241 thread T0
    #0 0x5649b384aa34 in dotzsptagwalk /home/hops/src/hcxtools/hcxpcaptool.c:4423:21
    #1 0x5649b384aa34 in processtzsppacket /home/hops/src/hcxtools/hcxpcaptool.c:4471
    #2 0x5649b384b7df in processudppacket /home/hops/src/hcxtools/hcxpcaptool.c:4530:2
    #3 0x5649b384c62a in processipv4packet /home/hops/src/hcxtools/hcxpcaptool.c:4724:2
    #4 0x5649b384e013 in processloopbackpacket /home/hops/src/hcxtools/hcxpcaptool.c:4993:2
    #5 0x5649b384e013 in processpacket /home/hops/src/hcxtools/hcxpcaptool.c:5031
    #6 0x5649b3851e1a in processpcap /home/hops/src/hcxtools/hcxpcaptool.c:5825:4
    #7 0x5649b3853ffb in processcapfile /home/hops/src/hcxtools/hcxpcaptool.c:6093:2
    #8 0x5649b385716c in main /home/hops/src/hcxtools/hcxpcaptool.c:6585:2
    #9 0x7fc05d3ccce2 in __libc_start_main (/usr/lib/libc.so.6+0x23ce2)
    #10 0x5649b370d24d in _start (/home/hops/src/hcxtools/hcxpcaptool+0x2a24d)

Address 0x7ffecf399241 is located in stack of thread T0 at offset 524449 in frame
    #0 0x5649b38517af in processpcap /home/hops/src/hcxtools/hcxpcaptool.c:5726

  This frame has 3 object(s):
    [32, 56) 'pcapfhdr' (line 5730)
    [96, 113) 'pcaprhdr' (line 5731)
    [160, 524448) 'packet' (line 5732) <== Memory access at offset 524449 overflows this variable
HINT: this may be a false positive if your program uses some custom stack unwind mechanism, swapcontext or vfork
      (longjmp and C++ exceptions *are* supported)
SUMMARY: AddressSanitizer: stack-buffer-overflow /home/hops/src/hcxtools/hcxpcaptool.c:4423:21 in dotzsptagwalk
Shadow bytes around the buggy address:
  0x100059e6b1f0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x100059e6b200: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x100059e6b210: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x100059e6b220: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x100059e6b230: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
=>0x100059e6b240: 00 00 00 00 00 00 00 00[f3]f3 f3 f3 f3 f3 f3 f3
  0x100059e6b250: f3 f3 f3 f3 f3 f3 f3 f3 f3 f3 f3 f3 f3 f3 f3 f3
  0x100059e6b260: f3 f3 f3 f3 f3 f3 f3 f3 00 00 00 00 00 00 00 00
  0x100059e6b270: 00 00 00 00 f1 f1 f1 f1 f8 f8 f8 f8 f8 f8 f8 f8
  0x100059e6b280: f2 f2 f2 f2 f8 f8 f8 f8 f8 f8 f8 f8 f2 f2 f2 f2
  0x100059e6b290: f8 f2 f8 f2 00 00 00 00 00 00 00 00 00 00 00 00
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07 
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
  Shadow gap:              cc
==12587==ABORTING
```

On a i7-6700K CPU (using 4 threads) I get now around speed of 100k testcases per second.
I've also looked into oss-fuzz, the technical stuff shouldn't be a problem. However they might reject a request `To be accepted to OSS-Fuzz, an open-source project must have a significant user base and/or be critical to the global IT infrastructure.`
Anyway, once the low hanging fruits are weeded out I will try to integrate it and make a pull request (both here and in oss-fuzz)

**Comment by ZerBea** [comment]
I noticed that behavior, too and fixed it already by latest commit:
https://github.com/ZerBea/hcxtools/commit/3ccc7b571115e5c0d6117d4c73b2c0525f75f919

$ hcxpcaptool -V loop.bin
reading from loop.bin
unsupported pcap version: 16
summary:                                        
file name........................: loop.bin
file type........................: pcap 0.0
file hardware information........: unknown
file os information..............: unknown
file application information.....: unknown
network type.....................: DLT_NULL (0)
endianness.......................: little endian
read errors......................: yes
packets inside...................: 0
skipped packets.(damaged)........: 0
packets with GPS data............: 0
packets with FCS.................: 0

**Comment by hops** [comment]
Found another stack overflow
```
1MOyoQAAoREtHDAp990PAAAAAAAAAAAAAAAA/wAfGx8eAAAAAAADAwIAAQBABTE4NNXV1dXV1TIy
yDEyNzc1Hx4AAAAAACMDAgAAAEABMQAAAAAAAwMCAAAARAEAAAChEDKhES0ckJCQkJD6kJD/AEQB
AQAAEiMDAgAAAEABMQAAAAAAAwMCAAAARAEAAAChEDKhES0ckJCQkJT6kJAAAEQBAQAAACMDAgAA
AEABMQAAAAAAAwMCAAAARAEAAAChEDKhES0ckJCQkJD6kJAAAEQBAQAAyRAyoRFYWFhYWFhYWFhY
WFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYSFhYWFhYWFhY1VhYWFhYWFhYWFhYWFhYWFhYWFhYWFhY
WFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYpwAAAwEDAQAAACMDAgAAAEABMQAA
AAAAAwMCAAAARAEAAAChEDKhES4ckJCQkJD6kJAAAEQBAQAAyRAyoRFYWFhYWFhYWFhYWFhYWFhY
WFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhY1VhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhY
WFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWFhYWAAAAwEDAQAAACMDAgAAAEABMQAAAAAAAwMC
AAAARAEAAAChEDKhES0ckJCQkJD6kJAAAEQBAQAAyRAyoRFYWFhYWAA=
```
```
$ ./hcxpcaptool -o /dev/null fuzz-testcases/crash.bin 

reading from crash.bin
=================================================================
==5265==ERROR: AddressSanitizer: stack-buffer-overflow on address 0x7ffec7b03100 at pc 0x55dbfd41fef2 bp 0x7ffec7a82f40 sp 0x7ffec7a82f30
READ of size 1 at 0x7ffec7b03100 thread T0
    #0 0x55dbfd41fef1 in dotzsptagwalk /home/hops/src/hcxtools/hcxpcaptool.c:4423
    #1 0x55dbfd41fef1 in processtzsppacket /home/hops/src/hcxtools/hcxpcaptool.c:4469
    #2 0x55dbfd420130 in processtzsppacket /home/hops/src/hcxtools/hcxpcaptool.c:4439
    #3 0x55dbfd420130 in processudppacket /home/hops/src/hcxtools/hcxpcaptool.c:4528
    #4 0x55dbfd4202c0 in processudppacket /home/hops/src/hcxtools/hcxpcaptool.c:4501
    #5 0x55dbfd4202c0 in processipv4packet /home/hops/src/hcxtools/hcxpcaptool.c:4722
    #6 0x55dbfd422038 in processipv4packet /home/hops/src/hcxtools/hcxpcaptool.c:4987
    #7 0x55dbfd422038 in processloopbackpacket /home/hops/src/hcxtools/hcxpcaptool.c:4991
    #8 0x55dbfd422038 in processpacket /home/hops/src/hcxtools/hcxpcaptool.c:5029
    #9 0x55dbfd424936 in processpcap /home/hops/src/hcxtools/hcxpcaptool.c:5835
    #10 0x55dbfd4256f9 in processcapfile /home/hops/src/hcxtools/hcxpcaptool.c:6103
    #11 0x55dbfd3ffd61 in main /home/hops/src/hcxtools/hcxpcaptool.c:6595
    #12 0x7fc44706fce2 in __libc_start_main (/usr/lib/libc.so.6+0x23ce2)
    #13 0x55dbfd40033d in _start (/home/hops/src/hcxtools/hcxpcaptool+0xa33d)

Address 0x7ffec7b03100 is located in stack of thread T0 at offset 524448 in frame
    #0 0x55dbfd42421f in processpcap /home/hops/src/hcxtools/hcxpcaptool.c:5730

  This frame has 3 object(s):
    [32, 56) 'pcapfhdr'
    [96, 113) 'pcaprhdr'
    [160, 524448) 'packet' <== Memory access at offset 524448 overflows this variable
HINT: this may be a false positive if your program uses some custom stack unwind mechanism or swapcontext
      (longjmp and C++ exceptions *are* supported)
SUMMARY: AddressSanitizer: stack-buffer-overflow /home/hops/src/hcxtools/hcxpcaptool.c:4423 in dotzsptagwalk
Shadow bytes around the buggy address:
  0x100058f585d0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x100058f585e0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x100058f585f0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x100058f58600: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x100058f58610: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
=>0x100058f58620:[f3]f3 f3 f3 00 00 00 00 00 00 00 00 00 00 00 00
  0x100058f58630: 00 00 00 00 f1 f1 f1 f1 f8 f2 f2 f2 f2 f2 f2 f2
  0x100058f58640: f8 f2 f2 f2 f2 f2 f2 f2 00 00 00 00 00 00 00 00
  0x100058f58650: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x100058f58660: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x100058f58670: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07 
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
==5265==ABORTING
```

**Comment by ZerBea** [comment]
Yes, noticed that, too and fixed it already by commit:
https://github.com/ZerBea/hcxtools/commit/4144839dea1f16ae366254c380c5fa7abd84395a

$ hcxpcaptool -V loop.bin

reading from loop.bin
failed to set file pointer
summary:                                        
file name........................: loop.bin
file type........................: pcap 0.4513
file hardware information........: unknown
file os information..............: unknown
file application information.....: unknown
network type.....................: DLT_NULL (0)
endianness.......................: little endian
read errors......................: yes
packets inside...................: 2
skipped packets.(damaged)........: 0
packets with GPS data............: 0
packets with FCS.................: 0
IPv4 packets.....................: 1
UDP packets......................: 1

**Comment by hops** [comment]
I alreadz tested it against that commit and it does trigger the bug when compiled with Address Sanitizer e.g.
`cc -O3 -Wall -Wextra -ggdb -std=gnu99 -fsanitize=address  -MMD -MF .deps/hcxpcaptool.d -o hcxpcaptool hcxpcaptool.c -lz -lcrypto`

**Comment by ZerBea** [comment]
ok, fixed, now:

cc -O3 -Wall -Wextra -ggdb -std=gnu99 -fsanitize=address -MMD -MF .deps/hcxpcaptool.d -o hcxpcaptool hcxpcaptool.c -lz -lcrypto

$ hcxpcaptool -V /home/zerobeat/temp/loop.pcapng

reading from loop.pcapng
unsupported pcap version: 0
summary:                                        
file name........................: loop.pcapng
file type........................: pcap 0.0
file hardware information........: unknown
file os information..............: unknown
file application information.....: unknown
network type.....................: DLT_NULL (0)
endianness.......................: little endian
read errors......................: yes
packets inside...................: 0
skipped packets.(damaged)........: 0
packets with GPS data............: 0
packets with FCS.................: 0

**Comment by hops** [comment]
The root cause doesn't seem to be fixed. Here's a tetscase that has pcap version 2 but still triggers the bug under Address Sanitizer
```
1MOyoQIAAAAAAAD3A308+Ym9gAAAAAAAAADlbIVSCx+0AAAAIMAAAAIAAABtGxERERERFhAQEBBx
NDFiEBAQtRAQEP//////////EBAQEBTvEBAQEBCQkBAAAAABAAAScXFxcXFxcXFxcXFxcXHdF4Jc
cCtRsdRXh07koavmCS6P4ciJ/3EyoTCIf1Wiau7LehmYgIzKJALKWqb4Cwww400skYzPP9tp60/s
54NJ0zAQhKhxm6rNlka7Olc2rxjqUvttUQVwcriCgQyRLwWTspw7jSTVwTAQzBrBoqAZKyRp+qGi
Dv25AKBkHw5xcXFxcXEQEBAQEJnER2ZMrMUMQqivzTD932biZXEE1Z49AsueNL7IoLdqRKTbIUzg
Uyg7szJWxSJNaUfGZE24Zty6BB4QEBAQLTk4MjgxMTYzODkxNjExMTY3cRAQEBAQEBAQEBAQEBCQ
jD82QS17c0rCCkRyhcmC5fYPrapxcXFxcXEQEBAQEJnER2ZMrOUMQqivzTD932biZXEE1Z49Asue
NL7IoLdqRKTbIUzgUyg7szJWxSJNaUfGZE24Zty6BB4QEBAQLTk4MjgxMTYzODkxNjExMTY3cRAQ
EBAQEBAQEBAQEBCQjD82QS17c0rCCkRyhcmC5fYPraolEBAQEC05ODI4MTE2Mzg5MTYxMTE2N3EQ
EBAQEBAQEBAQEBAQkIw/NkEte3NKwgpEcoXJguX2D62qJYr6XvQ0SYl776OO9KeLKorIw3FxcXFx
cXFxcXFxcXFxcXFxcXFxcXFxcXFxcXFxcXFxcXFxcXFxcXFxAA==
```

**Comment by ZerBea** [comment]
At least I found the issue and pushed a fix:
fixed detection of false IPv6 and IPv4 frame length insdie an UDP packet

$ cc -O3 -Wall -Wextra -ggdb -std=gnu99 -fsanitize=address -MMD -MF .deps/hcxpcaptool.d -o hcxpcaptool hcxpcaptool.c -lz -lcrypto

$ ./hcxpcaptool -V /home/zerobeat/temp/loop.pcapng

reading from loop.pcapng
failed to read packet 1          
summary:                                        
file name........................: loop.pcapng
file type........................: pcap 2.0
file hardware information........: unknown
file os information..............: unknown
file application information.....: unknown
network type.....................: DLT_NULL (0)
endianness.......................: little endian
read errors......................: yes
packets inside...................: 1
skipped packets.(damaged)........: 0
packets with GPS data............: 0
packets with FCS.................: 0

**Comment by hops** [comment]
Yes that fixed it!
After half an hour of fuzzing (testing 140 million cases) on the current master I only found two segfaults which I can't reproduce outside of honggfuzz. My theory is, it has to do with my additions to make the code work with libFuzzer/honggfuzz. Maybe I don't clean up correctly... I will have a closer look at this later.
I'll let the fuzzer run for a bit longer and create a code coverage report to see which parts of the code have been fuzzed.
Also, so far I only fuzzed hcxpcaptool but if you want to release version 5.1.7, go ahead, I most likely won't be able to have a look at the other tools today.

**Comment by ZerBea** [comment]
Due to the several stack smashing issues I decided to refactor the talk walk engine by this commit:
https://github.com/ZerBea/hcxtools/commit/55c49e52cf2fc2644978d878b005b3cdb672c3bc
Now we should be save - I  hope so.

**Comment by hops** [comment]
Your refactoring seem to have introduced new bugs, see [testcases.zip](https://github.com/ZerBea/hcxtools/files/3296061/testcases.zip)
There are three crashes that are most likely related since they crash in the same function.
All of them were not present in 66767ee

**Comment by ZerBea** [comment]
Fixed that issue by latest commit:
https://github.com/ZerBea/hcxtools/commit/0dc22d39c5e3d9f6399b77696a7659b5a23c7512
Now we check if the file pointer is behind the anonced block length before we start the pacpng option walk:
$ cc -O3 -Wall -Wextra -ggdb -std=gnu99 -fsanitize=address -MMD -MF .deps/hcxpcaptool.d -o hcxpcaptool hcxpcaptool.c -lz -lcrypto

$ ./hcxpcaptool -V crash1.bin
reading from crash1.bin
block length greater than file size
summary:                                        
file name........................: crash1.bin
file type........................: pcapng 1.0
file hardware information........: unknown
file os information..............: unknown
file application information.....: unknown
network type.....................: DLT_NULL (0)
endianness.......................: little endian
read errors......................: yes
packets inside...................: 0
skipped packets (damaged)........: 0
packets with GPS data............: 0
packets with FCS.................: 0


$ ./hcxpcaptool -V crash2.bin
reading from crash2.bin
block length greater than file size
summary:                                        
file name........................: crash2.bin
file type........................: pcapng 1.0
file hardware information........: unknown
file os information..............: unknown
file application information.....: unknown
network type.....................: DLT_NULL (0)
endianness.......................: little endian
read errors......................: yes
packets inside...................: 0
skipped packets (damaged)........: 0
packets with GPS data............: 0
packets with FCS.................: 0


$ ./hcxpcaptool -V crash3.bin
reading from crash3.bin
block length greater than file size
summary:                                        
file name........................: crash3.bin
file type........................: pcapng 1.0
file hardware information........: unknown
file os information..............: unknown
file application information.....: unknown
network type.....................: DLT_NULL (0)
endianness.......................: little endian
read errors......................: yes
packets inside...................: 0
skipped packets (damaged)........: 0
packets with GPS data............: 0
packets with FCS.................: 0

**Comment by RealEnder** [comment]
Two more from me, those were AFL founds for old version, which I missed to send to Mike, but trigger with current HEAD.
```
Cg0NCpUAAABNPSsaAQAAAP//CABG8gAA/wAAAJ3yJwDtwSaA6NCDha5Qh47///////8CAAYAYVNt
djdsAAADABQATGludXggNC4xNC42Ny0xLUFSBgAAAMQAaGN4ZHVtcHRvb2wgNC4yLjEAAACc8ggA
RvIAAAAAAOec8icA7cEmgAAAAOjQg4WuUIeOTh8mWkundMM8MJDch32a4gDLhW0ABgAAANAAAAEA
AAAsAAAAfwAAAP//AAACAAUAd2xhbjAAAADxHAAAAAAAAABzAAAALAAAfwIAAABkAAAAAP/xakFz
EEhgbC0abhED/wAAAAAAAAAAAAAAAP/xakFzEEhgbC0abhED/wAAAAAAAAAAAAAARAABARBHABDW
UKAMrIGEDAMAAABkAAAHBlVTIAELGyoBADAYAQAAD6wCAgACAAAyBDBIYGwtGm4RA/8AAAAAAAAA
AAAAAAAAAAAAAAAAAAAAPRYBBRUAAAAgAAAAAAAAAAAAAAAAAAAA3RgAUPICAQGDAAOkAAAnpACC
AQBeAGIyLwDdCQADfwEBAADyAlUAAFDyBBBOAAEQEEQAAQEQRwAQAAAAAAAAEA==
```
The other one I will send by email, since it's 86K.

**Comment by ZerBea** [comment]
I did a complete refactoring of pcapng option walk function and committed this, now:
https://github.com/ZerBea/hcxtools/commit/eeef44c158108e40552fe1aa093fccad0864a626
@RealEnder this includes your last crash, too.

At least we have now 1024 bytes for an option comment. If the comment is greater, we will reject it.
We do not process an error that cause hcxpcaptool stop reading date, we only disegard this comment.
Please test.....

**Comment by ZerBea** [comment]
pushed another fix.
Alow only up to 1023 bytes to read, 1024 is zero to terminate string!

**Comment by ZerBea** [comment]
The second crash test passes hcxpcaptool also. As expected we keep as much data as possible:
$ ./hcxpcaptool -V id:000001,sig:06,src:000002,op:flip1,pos:96
reading from id:000001,sig:06,src:000002,op:flip1,pos:96
summary:                                        
file name........................: id:000001,sig:06,src:000002,op:flip1,pos:96
file type........................: pcapng 1.0
file hardware information........: armv7l
file os information..............: Linux 4.14.67-1-ARCH
file application information.....: hcxdumptool 4.2.1
network type.....................: DLT_IEEE802_11_RADIO (127)
endianness.......................: little endian
read errors......................: flawless
packets inside...................: 488
skipped packets (damaged)........: 0
packets with GPS data............: 0
packets with FCS.................: 0
beacons (total)..................: 37
probe requests...................: 11
probe responses..................: 17
association requests.............: 5
association responses............: 9
reassociation requests...........: 1
reassociation responses..........: 2
authentications (OPEN SYSTEM)....: 172
authentications (BROADCOM).......: 172
EAPOL packets (total)............: 234
EAPOL packets (WPA1).............: 3
EAPOL packets (WPA2).............: 231
PMKIDs (total)...................: 22
PMKIDs (WPA1)....................: 3
PMKIDs (WPA2)....................: 208
PMKIDs from access points........: 22
best handshakes..................: 2 (ap-less: 0)
best PMKIDs......................: 3

**Comment by RealEnder** [comment]
Looked at hangs from last run against current HEAD. 67 left, from which there are at least 3 different hang types. Will send the cases also via e-mail.

**Comment by ZerBea** [comment]
Should be fixed by latest commit:
https://github.com/ZerBea/hcxtools/commit/2d349711504f512d6e23d1ede30cee4a44b5f6ba

**Comment by hops** [comment]
I figured out the crashes that weren't reproducible outside of honggfuzz and as expected I didn't reset a global variable.

Also I've turned on every possible output option (except GPS) and found a heap buffer overflow in the outputpppchaplist() function (only triggers with Address Sanitizer).
Except for this bug, hcxpcaptool looks pretty stable now. I haven't really touched the other tools but from a quick look hcxpcaptool seems to be the most complex. I haven't forgotten about the oss-fuzz integration but I won't have much time to work on it until Thursday. I think I'll open separate issue to have a better discussion.

```
1MOyoQIAAAAAAAEADAAAAEobHxcAAAAADQAAAAYIIxyfAAAAAAMDAwD8m8uddC0xMDY1NzQwNDYy
NjU4MTUzNjg0d+0DY20bM2/bkvQ3WBXH1Yst6TFev6gN6ULQnoStngwWn+lTcCkTTZv4aR6VzMmT
OOYkmcxxJO4khU/bDvUVsuxQKWeRzhq718WGAG/4685R2ehGX2uM6xG6MJ/+E/hZ56FzXmlzv937
Rse443Nnz/5hWq6xb5wvcrN0dnRzvl/OGxzZYhdsygAGThsfnwAAAACmp6cCAAAAZ6enAACAL07m
p6enp6enp6enp6enp6enp6enp6enp6eJiYmJiYmJiYmJiAvhxYgDWYkj6sIjAgAAEAEbDgAAAAAA
DABu5YbkxtjebqY/flUqcBkW9IryladTJkQupzbvwGVMdkkShiRWZ0WkR4U7X8yjxDOdhtjJdtNb
CICHIBax4RK5uEPIhabj+CANjILHZlgWrj3szy92ux4NAAAABk4bH58AAAAApqenAgAAAGdYpwAA
gC9O5qenp6enp6enp6enp6enp6enp6enp6eniYmJiYmJiYmJiYgL4cWIA1mJI+rCIwEAABAB0AAA
AAAAAAAAbuWG5MbY3m6mP35VKnAZFvSK0odfPFZx2RnqcSBdf7Z/HZHJm+esDhuiVSljwV0ptdgt
7y23fgnnQrRJz07qPJtF3SJ4bzAt0RATJZjqgJqIKLQJZihFfrXAHWFB+N7gIe9YB+i6L4TSCnL+
SWrxVEzHi4+XQRWymh7/XyTVWYDdSq8546ZlX7pyzAS0wdLyX2Ijrq2Uw9Qu4cYWlF3Kpxu+xPVK
TcGcQfi1pMkpQMwkYugk0BhNbTeu3h54BRyn99H9lDzo6baP00VEwoPKdSIkV0+1Cl7ZOmr7P17j
```
```
$ ./hcxpcaptool --netntlm-out=/dev/null crash.bin

reading from crash.bin
failed to read pcap packet header for packet 4
                                                
summary:                                        
--------
file name........................: crash.bin
file type........................: pcap 2.0
file hardware information........: unknown
file os information..............: unknown
file application information.....: unknown
network type.....................: DLT_NULL (0)
endianness.......................: little endian
read errors......................: yes
packets inside...................: 4
skipped packets (damaged)........: 0
packets with GPS data............: 0
packets with FCS.................: 0
IPv6 packets.....................: 2
GRE packets......................: 2
found............................: PPP-CHAP Authentication

=================================================================
==27457==ERROR: AddressSanitizer: heap-buffer-overflow on address 0x619000000075 at pc 0x563204102406 bp 0x7fff37d9c060 sp 0x7fff37d9c050
READ of size 1 at 0x619000000075 thread T0
    #0 0x563204102405 in fwritehexbuffraw include/fileops.c:114
    #1 0x563204102405 in outputpppchaplist /tmp/hcxtools-master/hcxpcaptool.c:2060
    #2 0x563204114817 in processcapfile /tmp/hcxtools-master/hcxpcaptool.c:6137
    #3 0x5632040eed91 in main /tmp/hcxtools-master/hcxpcaptool.c:6579
    #4 0x7fcf5934bce2 in __libc_start_main (/usr/lib/libc.so.6+0x23ce2)
    #5 0x5632040ef36d in _start (/tmp/hcxtools-master/hcxpcaptool+0xa36d)

0x619000000075 is located 11 bytes to the left of 1030-byte region [0x619000000080,0x619000000486)
allocated by thread T0 here:
    #0 0x7fcf59aca801 in __interceptor_realloc /build/gcc/src/gcc/libsanitizer/asan/asan_malloc_linux.cc:105
    #1 0x563204104005 in addpppchapleap /tmp/hcxtools-master/hcxpcaptool.c:2431

SUMMARY: AddressSanitizer: heap-buffer-overflow include/fileops.c:114 in fwritehexbuffraw
Shadow bytes around the buggy address:
  0x0c327fff7fb0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x0c327fff7fc0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x0c327fff7fd0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x0c327fff7fe0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x0c327fff7ff0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
=>0x0c327fff8000: fa fa fa fa fa fa fa fa fa fa fa fa fa fa[fa]fa
  0x0c327fff8010: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x0c327fff8020: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x0c327fff8030: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x0c327fff8040: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x0c327fff8050: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07 
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
==27457==ABORTING
```

**Comment by ZerBea** [comment]
Great, another one fixed by commit:
https://github.com/ZerBea/hcxtools/commit/cab4020c0a958ef0992a0f4d8e7ce34320d72753

Now running gcc 9.1.0:
 cc -O3 -Wall -Wextra -ggdb -std=gnu99 -fsanitize=address -MMD -MF .deps/hcxpcaptool.d -o hcxpcaptool hcxpcaptool.c -lz -lcrypto
$ ./hcxpcaptool --netntlm-out=/dev/null /home/zerobeat/temp/testcases/crash.bin
reading from crash.bin
failed to read packet 3          
summary:                                        
file name........................: crash.bin
file type........................: pcap 2.0
file hardware information........: unknown
file os information..............: unknown
file application information.....: unknown
network type.....................: DLT_NULL (0)
endianness.......................: little endian
read errors......................: yes
packets inside...................: 3
skipped packets (damaged)........: 0
packets with GPS data............: 0
packets with FCS.................: 0
IPv6 packets.....................: 2
GRE packets......................: 2

**Comment by ZerBea** [comment]
Yes, hcxpcaptool is really complex.
Tons of nested frames starting with pcapng -> radiotap -> 802.11 -> ... IPv6 -> UDP, for example.
Several different authentications (AKM) and many other features (upcomming device information evaluation, for example)
Fixed also several other issues (received from Alex) and I hope hcxpcaptool is bulletproof (@RealEnder: I like "your" word), now.

**Comment by hops** [comment]
The heap overflow can still be triggered. The testcase looks a bit weird because it's minimized with afl-tmin which replaces unrelated data with ASCII zeros.
```
1MOyoQIAMDAwMDAwMDAwMDAwMDAAAAAAMDAwMDAwMDCfAAAAMDAwMDAwMDAwMDAwMDAwMDAwMDAw
MDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAw
MDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAw
MDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwnwAAADAwMDAwMDAwMDAwMDAwMDAw
MDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAw
MDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAw
MDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMJ8AAAAwMDAwAgAAAGcwMDAA
gC8wMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwiYgLMDAwMDAwMDAwMDAwwiMCMAAw
FTAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAw
MDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDCfAAAAMDAwMAIAAABn
MDAwAIAvMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMImICzAwMDAwMDAwMDAwMMIj
ATAAcDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAw
MDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMDAwMA==
```
There's no easy way (I'm aware of) to create a code coverage report from fuzzing input. There is [afl-cov](https://github.com/mrash/afl-cov) but that seems to only work out of the box with AFL. Also after thinking about it I came to the conclusion that a code coverage reports might not be a very good measurement because it only tells that a line of code was hit but not how many times. For example after 1.4 billion tests with honggfuzz the heap overflow was only triggered seven times. So a code coverage report of course helps identifying parts of the code that hasn't been reached but there's no way to tell how thorough it has been fuzzed. Such a feature would have to be built into a fuzzer and I'm not aware of any that does it (I'd love to be wrong about that!).

**Comment by ZerBea** [comment]
Now we are running hard coded values of ppp chap authentication frames:
challenge 16 bytes
response 49 bytes
That seems to work by last commit:
https://github.com/ZerBea/hcxtools/commit/067252902c612f328eaa950e64d979859a3bfd83

$ ./hcxpcaptool --netntlm-out=/dev/null crash.bin
reading from new.pcapng
summary:                                        
file name........................: crash.bin
file type........................: pcap 2.12336
file hardware information........: unknown
file os information..............: unknown
file application information.....: unknown
network type.....................: DLT_NULL (0)
endianness.......................: little endian
read errors......................: flawless
packets inside...................: 4
skipped packets (damaged)........: 0
packets with GPS data............: 0
packets with FCS.................: 0
IPv6 packets.....................: 2
GRE packets......................: 2
found............................: PPP-CHAP Authentication

**Comment by hops** [comment]
The issue isn't resolved yet. Here's a testcase that still triggers the same bug.
```
1MOyoQIABABzpybDsksyNDM4NDIBAAAADzM4Njg0+4V0AAAAAEdGAAATUOUogXSpiQtpgAgATADG
uzUxqs4BLy46Vmty+ULAIrE3kweCZP4gKNz9tVPU01XMNowoa1kjODUyASiBdamIC2mACNP5uH5g
AAAOE8IjAjcyMhAA5wrIky4XxGkQ6sAIMiDVTpTDsksyNDM4NTIBAAAADzM4Njg0+4V0AAAAAEdG
AAATUOUogXSpiQtpgAgATADGuzUxqs4BLy46Vmty+ULAIrE3kweCZP4gKNz9tVPU01XMNowoa1kj
ODUyASiBdamIC2mACNP5uPCBkj8OE8IjATcyMjGn5wrIky4XxGkQ6sAIMiDVTpTDsksyNDM4NTIB
AAAADzM4Njg1+4V0AAAAAEdGAAATUOUogXSpiQtpgAgATADGuzUxqs4BAA==
```
I got curious and had a closer look.
In `addpppchapleap()` we make sure `count` can be either 16 or 49 with this check:
https://github.com/ZerBea/hcxtools/blob/b99a811b7e1aba4122a6a1cfe4b64888408e994d/hcxpcaptool.c#L2420
`count` gets assigned to `leap2liste->len` at a later point.
In `outputleaplist()` we write some data of the response in hex
https://github.com/ZerBea/hcxtools/blob/b99a811b7e1aba4122a6a1cfe4b64888408e994d/hcxpcaptool.c#L2073
To reach this branch `zeigerrs->code` has to be `EAP_CODE_RESP`
https://github.com/ZerBea/hcxtools/blob/b99a811b7e1aba4122a6a1cfe4b64888408e994d/hcxpcaptool.c#L2067
The problem is we can craft an input that ends up having `zeigerrs->len` of 16 and `zeigerrs->code` being `EAP_CODE_RESP`
This makes the reference passed into `fwritehexbuffraw()` pointing outside of the structure because of this part `&zeigerrs->data[zeigerrs->len -25]`

One solution would be to extend the check in `addpppchapleap()` to something like this
```diff
-if((count != 16) && (count != 49))
+if(((code == EAP_CODE_REQ) && (count != 16)) || ((code == EAP_CODE_RESP) && (count != 49)))
```
Though I don't have any knowledge about this protocol so I'm not sure if that would be the correct fix.

**Comment by ZerBea** [comment]
With the last fixes we resolved the hash length error. Unfortunately there was a username length error, too.
We detect a ppp chap authentication within an IpV4 -> GRE frame. Then we test the hash length (16 or 49 - that worked already ) and the username length (< 255).
If the username is > 255 we do not truncate the frame any longer and disregard it, now. 
We have to check if there are user names > 255 characters outside in the wildness. If so, we can expand the field size easily.

$ ./hcxpcaptool --netntlm-out=/dev/null /home/zerobeat/temp/testcases/crash.bin
reading from crash.bin
failed to read packet 2          
summary:                                        
file name........................: crash.bin
file type........................: pcap 2.4
file hardware information........: unknown
file os information..............: unknown
file application information.....: unknown
network type.....................: DLT_EN10MB (1)
endianness.......................: little endian
read errors......................: yes
packets inside...................: 2
skipped packets (damaged)........: 0
packets with GPS data............: 0
packets with FCS.................: 0
IPv4 packets.....................: 2
GRE packets......................: 2
found............................: PPP-CHAP Authentication

**Comment by ZerBea** [comment]
complete refactoring of pcapng engine by commit:
https://github.com/ZerBea/hcxtools/commit/4f944b057dbe407501e6b7abb143b8b54ffc94e3
fixed also a small bug regarding the EHPB in hcxdumptool.

**Comment by ZerBea** [comment]
Fixed another infinity loop when pcapng running empty because of crash of the capturing engine (000000000000000) by this commit:
https://github.com/ZerBea/hcxtools/commit/55d606145f7ca16363a1f15b1760bc7af745338b

**Comment by ZerBea** [comment]
added several pcpang integrity checks by commit
https://github.com/ZerBea/hcxtools/commit/75e8f00f7b5cd12954000c4007b24f9f2b1608d8
It is very hard to run test suites against such complex structures which contain tons of nested frames, but please test it. If no show stoppers occur, we can release 5.1.7

**Comment by ZerBea** [comment]
Checked wifite code and noticed that the loop is wanted.
This is the part within pmkid.py line 113 ... 140:

        # Start hcxdumptool
        t = Thread(target=self.dumptool_thread)
        t.start()

        # Repeatedly run pcaptool & check output for hash for self.target.essid

Closed this issue, because it isn't caused by hcxpcaptool.
